src/composables/useNavigationGuards.ts

```diff
@@ -1,18 +1,90 @@
 import NProgress from 'nprogress';
-import { useRouter } from 'vue-router';
+import { Router, useRouter } from 'vue-router';
+
+import { Network } from '@balancer-labs/sdk';
 
 import { useSidebar } from './useSidebar';
 import useVeBal from './useVeBAL';
+import {
+  networkFromSlug,
+  getSubdomain,
+  handleNetworkUrl,
+  networkId,
+  networkSlug,
+  appUrl,
+} from '@/composables/useNetwork';
 
 // Progress bar config
 NProgress.configure({ showSpinner: false });
 let delayedStartProgressBar;
 
+function setNetworkAndRefresh(networkId: string, url: string, router: Router) {
+  document.write('');
+  localStorage.setItem('networkId', networkId);
+  window.location.href = url;
+  router.go(0);
+}
+
 export default function useNavigationGuards() {
   const router = useRouter();
   const { setShowRedirectModal, isVeBalSupported } = useVeBal();
   const { setSidebarOpen } = useSidebar();
 
+  /**
+   * Subdomain redirects - e.g. "polygon.balancer.fi/".
+   */
+  router.beforeEach((to, from, next) => {
+    const subdomain = getSubdomain(window.location.host);
+    const subdomainNetwork = networkFromSlug(subdomain);
+    if (subdomainNetwork) {
+      setNetworkAndRefresh(
+        subdomainNetwork.toString(),
+        `${appUrl().replace(subdomain, 'app')}${
+          to.params.networkSlug ? '' : networkSlug
+        }${to.fullPath}`,
+        router
+      );
+    }
+    next();
+  });
+
+  /**
+   * Network url redirects - e.g. "app.balancer.fi/#/arbitrum".
+   */
+  router.beforeEach((to, from, next) => {
+    const urlNetwork = networkFromSlug(to.params.networkSlug?.toString() ?? '');
+    if (urlNetwork) {
+      const noNetworkChangeCallback = () => next();
+      const networkChangeCallback = (urlNetwork?: Network) => {
+        setNetworkAndRefresh(
+          (urlNetwork ?? '').toString(),
+          `/#${to.fullPath}`,
+          router
+        );
+      };
+      handleNetworkUrl(
+        to.params.networkSlug.toString(),
+        noNetworkChangeCallback,
+        networkChangeCallback
+      );
+    } else {
+      const nonNetworkedRoutes = [
+        '/',
+        '/terms-of-use',
+        '/privacy-policy',
+        '/cookies-policy',
+      ];
+      if (to.redirectedFrom || !nonNetworkedRoutes.includes(to.fullPath)) {
+        localStorage.setItem('networkId', networkId.value.toString());
+        router.push({
+          path: `/${networkSlug}${to.redirectedFrom?.fullPath ?? to.fullPath}`,
+        });
+      } else {
+        next();
+      }
+    }
+  });
```

Comment:
@Matthews3301 Could you explain specifically what redirect cases this is handling? Just in the comments below. I'm trying to understand it but it's not super clear.

Reply:
1. E.g. `xxx.fi/polygon/trade` -> if current network != polygon, reload page with right network.
2. E.g. `xxx.fi/trade`(previous format) -> unless it's one of nonNetworkedRoutes, redirect to `xxx.fi/<current_network>/trade`.